array_api_compat/common/_helpers.py

```diff
@@ -801,6 +802,257 @@ def size(x):
+                "Usage: either\n"
+                "    at(x, idx).set(value)\n"
+                "or\n"
+                "    at(x)[idx].set(value)\n"
+                "(same for all other methods)."
+            )
+        if mode != "promise_in_bounds" and not is_jax_array(self.x):
+            xp = array_namespace(self.x)
+            raise NotImplementedError(
+                f"mode='{mode!r}' is not supported for backend {xp.__name__}"
+            )
+
+        copy = _parse_copy_param(self.x, copy)
+
+        if copy and is_jax_array(self.x):
+            # Use JAX's at[]
+            at_ = self.x.at[self.idx]
+            args = (y, ) if y is not _undef else ()
+            return getattr(at_, at_op)(*args, mode=mode, **kwargs), None
+
+        # Emulate at[] behaviour for non-JAX arrays
+        x = self.x.copy() if copy else self.x
+        return None, x
+
+    def get(self, copy: bool | None = True, **kwargs):
+        """Return x[idx]. In addition to plain __getitem__, this allows ensuring
+        that the output is (not) a copy and kwargs are passed to the backend."""
+        # Special case when xp=numpy and idx is a fancy index
+        # If copy is not False, avoid an unnecessary double copy.
+        # if copy is forced to False, raise.
+        if (
+            is_numpy_array(self.x) 
+            and (
+                isinstance(self.idx, (list, tuple))
+                or (is_numpy_array(self.idx) and self.idx.dtype.kind in "biu")
+            )
+        ):
+            if copy is True:
+                copy = None
+            elif copy is False:
+                raise ValueError(
+                    "Indexing a numpy array with a fancy index always "
+                    "results in a copy"
+                )
+
+        res, x = self._common("get", copy=copy, **kwargs)
+        if res is not None:
+            return res
+        return x[self.idx]
+
+    def set(self, y, /, **kwargs):
+        """x[idx] = y"""
+        res, x = self._common("set", y, **kwargs)
+        if res is not None:
+            return res
+        x[self.idx] = y
+        return x
+
+    def apply(self, ufunc, /, **kwargs):
+        """ufunc.at(x, idx)"""
+        res, x = self._common("apply", ufunc, **kwargs)
+        if res is not None:
+            return res
+        ufunc.at(x, self.idx)
+        return x
+
+    def _iop(self, at_op: str, elwise_op: Callable[[Array, Array], Array], y: Array, **kwargs):
+        """x[idx] += y or equivalent in-place operation on a subset of x
+
+        which is the same as saying
+            x[idx] = x[idx] + y
+        Note that this is not the same as 
+            operator.iadd(x[idx], y)
+        Consider for example when x is a numpy array and idx is a fancy index, which
+        triggers a deep copy on __getitem__.
+        """
+        res, x = self._common(at_op, y, **kwargs)
+        if res is not None:
+            return res
+        x[self.idx] = elwise_op(x[self.idx], y)
```

Comment:
One issue here: there are different semantics between JAX and NumPy for repeated indices as this is currently implemented. For example:
```python
import numpy as np
x = np.ones(2)
i = np.array([0, 0])
x[i] *= 2
x  # array([2., 1.])
```
```python
import jax.numpy as jnp
x = jnp.ones(2)
x.at[i].mul(2)  # Array([4., 1.], dtype=float32)
```
We could fix this by dispatching the NumPy version to `np.multiply.at`, or maybe we could document that repeated indices lead to different behavior on different backends.

Reply:
Wouldn't that be a numpy-specific hack? In order to make it generic for other writeable, non-numpy arrays, we would need to amend the `__array_ufunc__` API to add `method="at"`?

I'd definitely prefer documenting that such a behaviour is undefined. 
I can't honestly think of any sane use case for repeated indices in updates?

Reply:
It’s useful for things like implementing `bincount` or `histogram` in a vectorized manner

Reply:
Caveat added